rdt/transformers/categorical.py

```diff
@@ -455,14 +457,39 @@ def _fit(self, data):
     def _transform(self, data):
         """Replace each category with its corresponding integer value.
 
+        If a category has not been seen before, a random value is assigned.
+        If no categories have been fitted (and `data` is not empty), raise a `ValueError`.
+
         Args:
             data (pandas.Series):
                 Data to transform.
 
         Returns:
-            numpy.ndarray:
+            pd.Series.
         """
-        return pd.Series(data).map(self.categories_to_values)
+        # Find the data categories that were not seen during fitting. Convert all nan's
+        # to np.nan to avoid having nan values with different ids.
+        unique_data = {np.nan if pd.isna(x) else x for x in pd.unique(data)}
+        categories = {np.nan if pd.isna(x) else x for x in self.categories_to_values.keys()}
+        unseen_categories = unique_data - categories
+        if unseen_categories:
+            if len(self.categories_to_values) == 0:
+                raise ValueError('No categories have been fitted.')
+
+            warn(
+                f'Warning: The data contains new categories {unseen_categories} that were not '
+                'seen in the original data. Assigning them random values. If you want to model '
+                'new categories, please fit the transformer again with the new data.'
+            )
+
+        mapped_data = pd.Series(data).map(self.categories_to_values)
```

Comment:
I think map takes a function. Why don't we just pass it in here and have the function return the value in the dict if it's present, and a random value otherwise? Do we need to apply two separate functions here?

Reply:
Unfortunately, yeah. The actual solution would probably be not to allow nan's as keys of a dictionary (since they don't necessarily have the same id, you can't actually get the nan value back, e.g. `d = {float('nan')}; d[float('nan')]` throws a keyerror).

So basically, we can't just pass your function because even if you fit on the `nans` (so they are present in the `self.categories_to_values` dict) when you try to `get` it thinks the `nan` is not there. There may be some shenanigans which allow you to access them without iterating through the whole dictionary (that was what I was trying to do in the `Untitled.ipynb` notebook which I accidentally submitted), but I couldn't figure out how.

So yeah, best I could do to keep the vectorization spirit was map the seen data first, like it was doing before, and then apply the random values where the nans are.